smart-contract-verifier/src/solidity/client.rs

```diff
@@ -0,0 +1,80 @@
+use super::compiler::SolidityCompiler;
+use crate::{compiler::Compilers, middleware::Middleware, verifier::Success};
+use std::sync::Arc;
+
+pub struct ClientBuilder {
+    compilers: Arc<Compilers<SolidityCompiler>>,
+    middleware_stack: Vec<Arc<dyn Middleware<Success>>>,
+}
+
+impl ClientBuilder {
+    /// Convenience method to initialize new client builder.
+    ///
+    /// If you need to keep a reference to the compilers after initialization, use [`new_arc`].
+    ///
+    /// [`new_arc`]: Self::new_arc
+    pub fn new(compilers: Compilers<SolidityCompiler>) -> Self {
+        Self::new_arc(Arc::new(compilers))
+    }
+
+    /// Initialize new client builder. [`new`] is more ergonomic if you don't need the `Arc`.
+    ///
+    /// [`new`]: Self::new
+    pub fn new_arc(compilers: Arc<Compilers<SolidityCompiler>>) -> Self {
+        Self {
+            compilers,
+            middleware_stack: vec![],
+        }
+    }
+
+    /// Convenience method to attach middleware.
+    ///
+    /// If you need to keep a reference to the middleware after attaching, use [`with_arc`].
+    ///
+    /// [`with_arc`]: Self::with_arc
+    pub fn with<M>(self, middleware: M) -> Self
+    where
+        M: Middleware<Success>,
+    {
+        self.with_arc(Arc::new(middleware))
+    }
+
+    /// Add middleware to the chain. [`with`] is more ergonomic if you don't need the `Arc`.
+    ///
+    /// [`with`]: Self::with
+    pub fn with_arc(mut self, middleware: Arc<dyn Middleware<Success>>) -> Self {
+        self.middleware_stack.push(middleware);
+        self
+    }
+
+    /// Returns a [`Client`] using this builder configuration.
+    pub fn build(self) -> Client {
+        Client::new(self.compilers, self.middleware_stack)
+    }
+}
+
+pub struct Client {
+    compilers: Arc<Compilers<SolidityCompiler>>,
+    middleware_stack: Box<[Arc<dyn Middleware<Success>>]>,
```

Comment:
imo it's better to have a single middleware
if somebody want several middlewares, he can create a separate middleware with an array of middlewares :D 
the idea is that you try to add some complex logic (arrays) in middleware, which can actually be different, like somebody might filter some things, or dynamically register middlewares or something. You can try to implement everything here, but you obviously fail, so in my experience it's better to have just one thing and if somebody wants an array - just implement your own middleware as array, or hashmap, or something

Reply:
imagine if middleware could return "ok/not ok" (like a filter) (just for an example). then you could implement an array which call every middleware and then checks if any returned "not ok". or you could call them each one by one while they return "ok" and stop on first "not ok". implementing it inside middleware allows both behaviors

Reply:
For now, I can see 2 main middlewares to implement: the one that submits verified contract's Abi to signature provider, and the one that submits verified contracts to Sourcify. It is quite easy to imagine that someone (we) may need to send something else to some other services, and it seems to be the most used case. Accepting an array of middlewares provides the ergonomic way to do that and removes the burden of implementing a separate structure for the most probable case from the user.

Reply:
Another solution is to accept the only one middleware, but also implement a separate structure (e.g., `Composition`) which could be used by users to combine several middlewares. That solution has several advantages for sure (such that we can get rid of builder for clients and significantly reduce amount of code; also we could write better tests with checking that middleware is actually called for the clients, and checking that all middlewares are processed for the `Composition` struct.

But for me the structure with a client with several middlewares is more common and more expected (you can find it in [`actix`](https://actix.rs/), [`reqwest_middleware`](https://docs.rs/reqwest-middleware/latest/reqwest_middleware/), and many more)

Reply:
I'm still not sure that it is better than the single middleware with `Composition` ability, though